File: src/runtime/dns_jsc/options_jsc.rs
//! `from_js`/`to_js` for `GetAddrInfo` and its nested option types, plus
//! `address_to_js`/`addr_info_to_js_array`. The pure types stay in `src/dns/`.

use bun_jsc::{
    ComptimeStringMapExt as _, JSGlobalObject, JSValue, JsError, JsResult, StringJsc as _,
};

use bun_dns::OptionsFromJsError as Invalid;
use bun_dns::{
    BACKEND_LABEL, Backend, FAMILY_MAP, Family, GetAddrInfoResult as GaiResult, Options,
    PROTOCOL_MAP, Protocol, ResultAny, SOCKET_TYPE_MAP, SocketType,
};
use bun_dns::{addr_info_count, address_to_string};

/// From-JS parse failure for `GetAddrInfo` options: a real pending JS error,
/// or a validation failure (`bun_dns::OptionsFromJsError`, which is JSC-free).
pub(crate) enum FromJSError {
    Js(JsError),
    Invalid(Invalid),
}
impl From<JsError> for FromJSError {
    fn from(e: JsError) -> Self {
        Self::Js(e)
    }
}
impl From<Invalid> for FromJSError {
    fn from(e: Invalid) -> Self {
        Self::Invalid(e)
    }
}

pub(crate) fn options_from_js(
    value: JSValue,
    global: &JSGlobalObject,
) -> Result<Options, FromJSError> {
    if value.is_empty_or_undefined_or_null() {
        return Ok(Options::default());
    }

    if value.is_object() {
        let mut options = Options::default();

        if let Some(family) = value.get(global, "family")? {
            options.family = family_from_js(family, global)?;
        }

        if let Some(socktype) = match value.get(global, "socketType")? {
            some @ Some(_) => some,
            None => value.get(global, "socktype")?,
        } {
            options.socktype = socket_type_from_js(socktype, global)?;
        }

        if let Some(protocol) = value.get(global, "protocol")? {
            options.protocol = protocol_from_js(protocol, global)?;
        }

        if let Some(backend) = value.get(global, "backend")? {
            options.backend = backend_from_js(backend, global)?;
        }

        if let Some(flags) = value.get(global, "flags")? {
            if !flags.is_number() {
                return Err(Invalid::InvalidFlags.into());
            }

            // Coerce to i32 and store/bit-test as u32.
            let flags_int: i32 = flags.coerce::<i32>(global)?;
            options.flags = flags_int;

            // hints & ~(AI_ADDRCONFIG | AI_ALL | AI_V4MAPPED)) !== 0
            let filter: u32 =
                !((bun_dns::AI_ALL | bun_dns::AI_ADDRCONFIG | bun_dns::AI_V4MAPPED) as u32);
            let int: u32 = flags_int as u32;
            if int & filter != 0 {
                return Err(Invalid::InvalidFlags.into());
            }
        }

        return Ok(options);
    }

    Err(Invalid::InvalidOptions.into())
}

pub(crate) fn family_from_js(
    value: JSValue,
    global: &JSGlobalObject,
) -> Result<Family, FromJSError> {
    if value.is_empty_or_undefined_or_null() {
        return Ok(Family::Unspecified);
    }

    if value.is_number() {
        return match value.coerce::<i32>(global)? {
            0 => Ok(Family::Unspecified),
            4 => Ok(Family::Inet),
            6 => Ok(Family::Inet6),
            _ => Err(Invalid::InvalidFamily.into()),
        };
    }

    if value.is_string() {
        // `Family.map` is a `ComptimeStringMap` ported as
        // `bun_dns::FAMILY_MAP` (a `comptime_string_map!`); `.from_js` comes
        // from `bun_jsc::ComptimeStringMapExt`.
        return match FAMILY_MAP.from_js(global, value)? {
            Some(f) => Ok(f),
            None => {
                if value.to_js_string(global)?.length() == 0 {
                    return Ok(Family::Unspecified);
                }
                Err(Invalid::InvalidFamily.into())
            }
        };
    }

    Err(Invalid::InvalidFamily.into())
}

pub(crate) fn socket_type_from_js(
    value: JSValue,
    global: &JSGlobalObject,
) -> Result<SocketType, FromJSError> {
    if value.is_empty_or_undefined_or_null() {
        // Default to .stream
        return Ok(SocketType::Stream);
    }

    if value.is_number() {
        return match value.to_int32() {
            0 => Ok(SocketType::Unspecified),
            1 => Ok(SocketType::Stream),
            2 => Ok(SocketType::Dgram),
            _ => Err(Invalid::InvalidSocketType.into()),
        };
    }

    if value.is_string() {
        return match SOCKET_TYPE_MAP.from_js(global, value)? {
            Some(s) => Ok(s),
            None => {
                if value.to_js_string(global)?.length() == 0 {
                    return Ok(SocketType::Unspecified);
                }
                Err(Invalid::InvalidSocketType.into())
            }
        };
    }

    Err(Invalid::InvalidSocketType.into())
}

pub(crate) fn protocol_from_js(
    value: JSValue,
    global: &JSGlobalObject,
) -> Result<Protocol, FromJSError> {
    if value.is_empty_or_undefined_or_null() {
        return Ok(Protocol::Unspecified);
    }

    if value.is_number() {
        return match value.to_int32() {
            0 => Ok(Protocol::Unspecified),
            6 => Ok(Protocol::Tcp),
            17 => Ok(Protocol::Udp),
            _ => Err(Invalid::InvalidProtocol.into()),
        };
    }

    if value.is_string() {
        return match PROTOCOL_MAP.from_js(global, value)? {
            Some(p) => Ok(p),
            None => {
                let str = value.to_js_string(global)?;
                if str.length() == 0 {
                    return Ok(Protocol::Unspecified);
                }
                Err(Invalid::InvalidProtocol.into())
            }
        };
    }

    Err(Invalid::InvalidProtocol.into())
}

pub(crate) fn backend_from_js(
    value: JSValue,
    global: &JSGlobalObject,
) -> Result<Backend, FromJSError> {
    if value.is_empty_or_undefined_or_null() {
        return Ok(Backend::default());
    }

    if value.is_string() {
        return match BACKEND_LABEL.from_js(global, value)? {
            Some(b) => Ok(b),
            None => {
                if value.to_js_string(global)?.length() == 0 {
                    return Ok(Backend::default());
                }
                Err(Invalid::InvalidBackend.into())
            }
        };
    }

    Err(Invalid::InvalidBackend.into())
}

pub(crate) fn result_any_to_js(
    this: &ResultAny,
    global: &JSGlobalObject,
) -> JsResult<Option<JSValue>> {
    Ok(match this {
        ResultAny::Addrinfo(addrinfo) => {
            // LIFETIMES.tsv: GetAddrInfo.Result.Any.addrinfo is FFI → *mut libc::addrinfo
            // (nullable raw pointer, no Option wrapper).
            let addrinfo: *mut super::netc::addrinfo = *addrinfo;
            if addrinfo.is_null() {
                return Ok(None);
            }
            // SAFETY: addrinfo is a non-null *mut libc::addrinfo owned by the
            // resolver; valid for the duration of this call.
            Some(addr_info_to_js_array(unsafe { &*addrinfo }, global)?)
        }
        ResultAny::List(list) => 'brk: {
            let array = JSValue::create_empty_array(global, list.len())?;
            let items: &[GaiResult] = list.as_slice();
            for (i, item) in (0_u32..).zip(items.iter()) {
                array.put_index(global, i, result_to_js(item, global)?)?;
            }
            break 'brk Some(array);
        }
    })
}

pub(crate) fn result_to_js(this: &GaiResult, global: &JSGlobalObject) -> JsResult<JSValue> {
    let obj = JSValue::create_empty_object(global, 3);
    obj.put(global, b"address", address_to_js(&this.address, global)?);
    obj.put(
        global,
        b"family",
        // `bun_sys::net::Address` exposes `.family() -> i32`.
        match this.address.family() {
            f if f == super::netc::AF_INET as _ => JSValue::js_number(4.0),
            f if f == super::netc::AF_INET6 as _ => JSValue::js_number(6.0),
            _ => JSValue::js_number(0.0),
        },
    );
    obj.put(global, b"ttl", JSValue::js_number(f64::from(this.ttl)));
    Ok(obj)
}

pub(crate) fn address_to_js(
    // `bun_dns::Address` is the `bun_sys::net::Address` sockaddr wrapper.
    address: &bun_dns::Address,
    global: &JSGlobalObject,
) -> JsResult<JSValue> {
    let mut str = match address_to_string(address) {
        Ok(s) => s,
        Err(_) => return Err(global.throw_out_of_memory()),
    };
    str.transfer_to_js(global)
}

pub(crate) fn addr_info_to_js_array(
    addr_info: &super::netc::addrinfo,
    global: &JSGlobalObject,
) -> JsResult<JSValue> {
    let array = JSValue::create_empty_array(global, addr_info_count(addr_info) as usize)?;

    {
        let mut j: u32 = 0;
        let mut current: *const super::netc::addrinfo = addr_info;
        // SAFETY: `current` walks the getaddrinfo(3) singly-linked result list;
        // each node and its `ai_next` are valid until freeaddrinfo is called by
        // the owner (which outlives this call).
        while let Some(this_node) = unsafe { current.as_ref() } {
            if let Some(result) = GaiResult::from_addr_info(this_node) {
                array.put_index(global, j, result_to_js(&result, global)?)?;
                j += 1;
            }
            current = this_node.ai_next;
        }
    }

    Ok(array)
}

File: src/runtime/dns_jsc/options_jsc.zig
//! `fromJS`/`toJS` for `GetAddrInfo` and its nested option types, plus
//! `addressToJS`/`addrInfoToJSArray`. The pure types stay in `src/dns/`.

pub fn optionsFromJS(value: jsc.JSValue, globalObject: *jsc.JSGlobalObject) FromJSError!Options {
    if (value.isEmptyOrUndefinedOrNull())
        return Options{};

    if (value.isObject()) {
        var options = Options{};

        if (try value.get(globalObject, "family")) |family| {
            options.family = try Family.fromJS(family, globalObject);
        }

        if (try value.get(globalObject, "socketType") orelse try value.get(globalObject, "socktype")) |socktype| {
            options.socktype = try SocketType.fromJS(socktype, globalObject);
        }

        if (try value.get(globalObject, "protocol")) |protocol| {
            options.protocol = try Protocol.fromJS(protocol, globalObject);
        }

        if (try value.get(globalObject, "backend")) |backend| {
            options.backend = try Backend.fromJS(backend, globalObject);
        }

        if (try value.get(globalObject, "flags")) |flags| {
            if (!flags.isNumber())
                return error.InvalidFlags;

            options.flags = try flags.coerce(std.c.AI, globalObject);

            // hints & ~(AI_ADDRCONFIG | AI_ALL | AI_V4MAPPED)) !== 0
            const filter = ~@as(u32, @bitCast(std.c.AI{ .ALL = true, .ADDRCONFIG = true, .V4MAPPED = true }));
            const int = @as(u32, @bitCast(options.flags));
            if (int & filter != 0) return error.InvalidFlags;
        }

        return options;
    }

    return error.InvalidOptions;
}
pub fn familyFromJS(value: jsc.JSValue, globalObject: *jsc.JSGlobalObject) FromJSError!Family {
    if (value.isEmptyOrUndefinedOrNull())
        return .unspecified;

    if (value.isNumber()) {
        return switch (try value.coerce(i32, globalObject)) {
            0 => .unspecified,
            4 => .inet,
            6 => .inet6,
            else => return error.InvalidFamily,
        };
    }

    if (value.isString()) {
        return try Family.map.fromJS(globalObject, value) orelse {
            if ((try value.toJSString(globalObject)).length() == 0) {
                return .unspecified;
            }

            return error.InvalidFamily;
        };
    }

    return error.InvalidFamily;
}
pub fn socketTypeFromJS(value: jsc.JSValue, globalObject: *jsc.JSGlobalObject) FromJSError!SocketType {
    if (value.isEmptyOrUndefinedOrNull())
        // Default to .stream
        return .stream;

    if (value.isNumber()) {
        return switch (value.to(i32)) {
            0 => .unspecified,
            1 => .stream,
            2 => .dgram,
            else => return error.InvalidSocketType,
        };
    }

    if (value.isString()) {
        return try SocketType.map.fromJS(globalObject, value) orelse {
            if ((try value.toJSString(globalObject)).length() == 0)
                return .unspecified;

            return error.InvalidSocketType;
        };
    }

    return error.InvalidSocketType;
}
pub fn protocolFromJS(value: jsc.JSValue, globalObject: *jsc.JSGlobalObject) FromJSError!Protocol {
    if (value.isEmptyOrUndefinedOrNull())
        return .unspecified;

    if (value.isNumber()) {
        return switch (value.to(i32)) {
            0 => .unspecified,
            6 => .tcp,
            17 => .udp,
            else => return error.InvalidProtocol,
        };
    }

    if (value.isString()) {
        return try Protocol.map.fromJS(globalObject, value) orelse {
            const str = try value.toJSString(globalObject);
            if (str.length() == 0)
                return .unspecified;

            return error.InvalidProtocol;
        };
    }

    return error.InvalidProtocol;
}
pub fn backendFromJS(value: jsc.JSValue, globalObject: *jsc.JSGlobalObject) Backend.FromJSError!Backend {
    if (value.isEmptyOrUndefinedOrNull())
        return Backend.default;

    if (value.isString()) {
        return try Backend.label.fromJS(globalObject, value) orelse {
            if ((try value.toJSString(globalObject)).length() == 0) {
                return Backend.default;
            }

            return error.InvalidBackend;
        };
    }

    return error.InvalidBackend;
}
pub fn resultAnyToJS(this: *const Result.Any, globalThis: *jsc.JSGlobalObject) bun.JSError!?jsc.JSValue {
    return switch (this.*) {
        .addrinfo => |addrinfo| try addrInfoToJSArray(addrinfo orelse return null, globalThis),
        .list => |list| brk: {
            const array = try jsc.JSValue.createEmptyArray(globalThis, @as(u32, @truncate(list.items.len)));
            var i: u32 = 0;
            const items: []const Result = list.items;
            for (items) |item| {
                try array.putIndex(globalThis, i, try item.toJS(globalThis));
                i += 1;
            }
            break :brk array;
        },
    };
}
pub fn resultToJS(this: *const Result, globalThis: *jsc.JSGlobalObject) bun.JSError!JSValue {
    const obj = jsc.JSValue.createEmptyObject(globalThis, 3);
    obj.put(globalThis, jsc.ZigString.static("address"), try addressToJS(&this.address, globalThis));
    obj.put(globalThis, jsc.ZigString.static("family"), switch (this.address.any.family) {
        std.posix.AF.INET => JSValue.jsNumber(4),
        std.posix.AF.INET6 => JSValue.jsNumber(6),
        else => JSValue.jsNumber(0),
    });
    obj.put(globalThis, jsc.ZigString.static("ttl"), JSValue.jsNumber(this.ttl));
    return obj;
}
pub fn addressToJS(address: *const std.net.Address, globalThis: *jsc.JSGlobalObject) bun.JSError!jsc.JSValue {
    var str = addressToString(address) catch return globalThis.throwOutOfMemory();
    return str.transferToJS(globalThis);
}
pub fn addrInfoToJSArray(addr_info: *std.c.addrinfo, globalThis: *jsc.JSGlobalObject) bun.JSError!jsc.JSValue {
    const array = try jsc.JSValue.createEmptyArray(
        globalThis,
        addrInfoCount(addr_info),
    );

    {
        var j: u32 = 0;
        var current: ?*std.c.addrinfo = addr_info;
        while (current) |this_node| : (current = current.?.next) {
            try array.putIndex(
                globalThis,
                j,
                try GetAddrInfo.Result.toJS(
                    &(GetAddrInfo.Result.fromAddrInfo(this_node) orelse continue),
                    globalThis,
                ),
            );
            j += 1;
        }
    }

    return array;
}

const std = @import("std");

const bun = @import("bun");
const JSError = bun.JSError;

const addrInfoCount = bun.dns.addrInfoCount;
const addressToString = bun.dns.addressToString;

const GetAddrInfo = bun.dns.GetAddrInfo;
const Backend = GetAddrInfo.Backend;
const Family = GetAddrInfo.Family;
const Protocol = GetAddrInfo.Protocol;
const Result = GetAddrInfo.Result;
const SocketType = GetAddrInfo.SocketType;

const Options = GetAddrInfo.Options;
const FromJSError = Options.FromJSError;

const jsc = bun.jsc;
const JSValue = jsc.JSValue;
